### PR TITLE
Vladimir.shtarev/qd 15655master

### PR DESCRIPTION
**Pull request by DarkMechanikum**
# Pull Request Details

Added custom-plugins.zip caching.

## Description

Native macOS runs used to re-download custom plugins every time. The branch caches them under the version-branch dir, keyed by the plugins URL (build number), and skips the download when the cache is valid (non-empty custom-plugins + matching source marker).

Updates stage into a temp dir and swap in only on success. Failed downloads or removal problems keep/roll back the previous install, leave the old URL marker, and log a warning. Covered by unit/integration tests for cache hit/miss, URL change, failure/rollback, and related edge cases.

## Related Issue

[QD-15655](https://youtrack.jetbrains.com/issue/QD-15655)

## Motivation and Context

Re-download of custom plugins eats into performance

## How Has This Been Tested
Environment: macOS (darwin), local Go toolchain; Docker not required for these tests.

Automated

go test ./internal/core/startup/ -run 'TestIsCustomPluginsCacheValid|TestGetPluginsURL|TestDownloadCustomPlugins' -count=1 -v
go test ./internal/core/startup/ ./internal/core/ ./internal/platform/product/ -count=1
Coverage check on the custom-plugins helpers via -coverprofile (happy path + failure/rollback branches)
What the tests cover

Cache hit / miss, URL change re-download, empty or non-dir custom-plugins
Failed download keeps previous cache and logs “keeping previously cached plugins”
First-install failure logging; corrupt archive failure
Stale staging cleanup; abort when leftover backup can’t be removed; rollback when post-swap removal fails (URL marker unchanged)
getPluginsURL for .sit, -aarch64.sit, .win.zip, .tar.gz

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [**CONTRIBUTING**](https://github.com/JetBrains/qodana-cli/blob/master/CONTRIBUTING.md) document.
- [ ] My code follows the code style of this project.
- [ ] My commit messages are styled with [gitmoji](https://gitmoji.dev)
- [ ] My change requires a change to the [documentation](https://jetbrains.com/help/qodana).
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by bonepl — APPROVED** [review]
LGTM

File: internal/core/startup/installers.go [hunk 1]
```diff
@@ -371,34 +371,170 @@ func verifySha256(checksumFile string, checkSumUrl string, filePath string) {
...
-	err := utils.DownloadFile(archivePath, pluginsUrl, getInternalAuth(), spinner)
+	archivePath := filepath.Join(stagingDir, "custom-plugins.zip")
+	err = utils.DownloadFile(archivePath, pluginsUrl, getInternalAuth(), spinner)
 	if err != nil {
-		return fmt.Errorf("error while downloading plugins: %v", err)
+		return failedCustomPluginsUpdate(pluginsUrl, hadPrevious, fmt.Errorf("error while downloading plugins: %v", err))
 	}
 
-	_, err = fexec.Exec(".", "tar", "-xf", archivePath, "-C", targetDir)
+	_, err = fexec.Exec(".", "tar", "-xf", archivePath, "-C", stagingDir)
```
[the remark this answers is not in the document]
Comment (DarkMechanikum) [reply internal/core/startup/installers.go]: 👍
